power_grid_model_c/power_grid_model/include/power_grid_model/auxiliary/input.hpp

```diff
@@ -21,36 +21,78 @@ struct BaseInput {
     ID id;  // ID of the object
 };
 
-struct NodeInput : BaseInput {
+static_assert(std::is_standard_layout_v<BaseInput>);
+
+struct NodeInput {
+    ID id;  // ID of the object
     double u_rated;  // rated line-line voltage
+
+    // implicit conversions to BaseInput
+    operator BaseInput&() { return reinterpret_cast<BaseInput&>(*this); }
+    operator BaseInput const&() const { return reinterpret_cast<BaseInput const&>(*this); }
 };
 
-struct BranchInput : BaseInput {
+static_assert(std::is_standard_layout_v<NodeInput>);
```

Comment:
should i also check that the pointer to all base attribute members and/or alignment are identical? e.g.

```suggestion
static_assert(std::is_standard_layout_v<NodeInput>);
static_assert(std::alignment_of_v<NodeInput> == std::alignment_of_v<BaseInput>);
static_assert(offsetof(NodeInput, id) == offsetof(BaseInput, id));
```

Reply:
It should be check on `>=`. As long as the alignment of "child" class is bigger or equal than de "parent" class, it should be fine.

Reply:
assert of `offsetof` is good.

Reply:
@mgovers  is this comment resolved and the PR is ready for review again?

Reply:
no not yet, i've implemented, but i'm still struggling to find a good way to deal with the template case without getting 1000000 static checks per component type and it makes the files unreadable.

i can also add those static asserts to a separate file, e.g. `input_static_checks.hpp` that is then included in `update.hpp` so as to not affect the readability of the dataset declarations

any thoughts?

Reply:
put into separate directory `<...>/auxiliary/static_asserts/<...>.hpp`